core/src/main/scala/tech/beshu/ror/accesscontrol/factory/decoders/rules/GroupsRuleDecoder.scala

```diff
@@ -34,9 +34,23 @@ import tech.beshu.ror.accesscontrol.show.logs._
 import tech.beshu.ror.accesscontrol.utils.CirceOps._
 
 class GroupsRuleDecoder(usersDefinitions: Definitions[UserDef],
-                        implicit val caseMappingEquality: UserIdCaseMappingEquality)
+                        override implicit val caseMappingEquality: UserIdCaseMappingEquality) extends AbstractGroupsRuleDecoder(usersDefinitions, caseMappingEquality) {
+  override def createRule(settings: AbstractGroupsRule.Settings, caseMappingEquality: UserIdCaseMappingEquality) =
+    new GroupsRule(settings, caseMappingEquality)
+}
+
+class GroupsAndRuleDecoder(usersDefinitions: Definitions[UserDef],
+                           override implicit val caseMappingEquality: UserIdCaseMappingEquality) extends AbstractGroupsRuleDecoder(usersDefinitions, caseMappingEquality) {
+  override def createRule(settings: AbstractGroupsRule.Settings, caseMappingEquality: UserIdCaseMappingEquality) =
+    new GroupsAndRule(settings, caseMappingEquality)
+}
+
+abstract class AbstractGroupsRuleDecoder(usersDefinitions: Definitions[UserDef],
```

Comment:
1. it looks like it can be private
2. we can rename it to `BaseGroupsRuleDecoder` also here

Reply:
1. the abstract class cannot become private, doesn't compile. 
2. renamed 👍

Reply:
ah, right.